### ISSUE TITLE
Compile minihaskell output to WASM

### ISSUE THREAD

**Issue by paulcadman** [heliaxdev/minijuvix#50]
See https://hackmd.io/A72oR4wYR-msPuFmU8f5rg?view

We want to see if we can achieve the following goal using the minihaskell backend (as opposed to going via LLVM).

> Validity predicates
>
> The first few VPs can be very simple. These neeed to be able to compile through LLVM to WASM and be compatible with the validity predicate interface of the Anoma ledger.
>
> For background on the validity predicate interface of the Anoma ledger, see [vp_template](https://github.com/anoma/anoma/blob/master/wasm/vp_template/src/lib.rs) for the interface a validity predicate needs to provide - we’ll need to have this validate_tx interface provided by the WASM output of MiniJuvix. See [here](https://github.com/anoma/anoma/blob/master/vm_env/src/imports.rs#L584) for some functions that should be available (as extern in > MiniJuvix) to validity predicates written in the MiniJuvix frontend, which then need to be compiled to appropriate calls in the LLVM/WASM (they are already defined using the C FFI, so hopefully this should be straightforward to implement).

For this issue we will see if we compile the minihaskell output to WASM, including the aroma FFI calls.

**Comment by paulcadman** [comment]
We tried using the [asterius](https://github.com/tweag/asterius) Haskell to WebAssembly compiler.

A validity predicate must be made available to the WASM runtime as a top-level function with a particular type-signature. So we need to be able to export Haskell functions via FFI using Asteruius.

We tried using `foreign export javascript` as described in [tweag blog](https://www.tweag.io/blog/2018-09-12-asterius-ffi/) and in the [Asterius test case](https://github.com/tweag/asterius/blob/641ade855c95818527f1726cc58ea7b49befede1/asterius/test/jsffi/jsffi.hs). But this did not make the exported symbol available in the WASM output.

There's an [open issue from 2020](https://github.com/tweag/asterius/issues/686) for the same issue with exporting Haskell symbols that has not been looked at by maintainers of Asterius.

It seems like Asterius is not maintained. For example the [project roadmap](https://asterius.netlify.app/roadmap.html) has not been updated since Q3 2021.

In summary we didn't find a way to get the FFI to work and so we can't use Asterius to compile validity predicates for use on the Anoma ledger.
